### PR TITLE
codegen with new eq (aka. sast)

### LINKED ISSUE #86: codegen eqs with deps as functions, per offline design discussion
**tl;dr** @tiancizhong and i discussed/sketched how to codegen an **eqs' deps** as of _(see #56)_

... not sure if everyone else is up to date. whoever takes on this issue _(or is interested in it)_, if you're not sure what to do, ping here and i'll fill this in with a better description.

At the least, fixing this bug requires getting tests skipped in 3699e0c7a031c60 to pass:
- [ ] **low hanging fruit**: `tests/test-find-decl-assignment.eq`

  is actually just broken codegen test case i discovered while working on equation dependencies. totally unrelated to equation deps.
- [ ] `tests/test-lazyresolved-vars-increment-self.eq`
- [ ] `tests/test-lazyresolved-vars.eq`
- [ ] `tests/test-global-variable-overwrite.eq`
- [ ] `tests/test-global-variable`

### PR DESCRIPTION
**Pull request by nnhoang**
fixes  #86

### CODE REVIEW

File: src/tests/test-dependency-resolved-inside-function.eq [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+  a = {
+    dep = 10;
+    dep + 1;
+  }
+}
+
+MyCtx: find a {
+  print("a = %0.f\n", a);
+}
+
```
Comment (nnhoang) [inline src/tests/test-dependency-resolved-inside-function.eq]: @jzacsh I think this test breaks our `sast`. `a` is both in `deps` and `indeps`, which is not supposed to happen. Could you check where flaw is?
```
MyCtx : {
    deps:
        a: [dep]

    indeps:
        a : ```
            dep = 10.;
            dep + 1.;

        ```

    finds:
        "find_MyCtx_0": {
            [0]: {
                deps:
                    a: [dep]

                indeps:
                    a : ```
                        dep = 10.;
                        dep + 1.;

                    ```

            }

        }


}
```

Comment (jzacsh) [inline src/tests/test-dependency-resolved-inside-function.eq]: fyi: working on this in [`jzacsh-issue91-depstwoplaces`](https://github.com/rxie25/PLT2016Spring/compare/jzacsh-issue91-depstwoplaces) branch

**Comment by nnhoang** [comment]
@jzacsh if there is no bug, then this this pull request should pass all current tests after merging with yours

File: src/tests/test-dependency-resolved-inside-function.eq (hunk 1 above)
Comment (jzacsh) [inline src/tests/test-dependency-resolved-inside-function.eq]: fyi, i opened pull request #93 which solves this. once reviewed & merged, you should be unblocked

**Comment by jzacsh** [comment]
let me know if you have any questions now that #92 is fixed

File: src/codegen.ml [hunk 2]
```diff
@@ -97,48 +122,97 @@ let translate sast =
...
   in
 
-  let gen_function_call_in_find ctx =
-    let rec gen_function_call_for_multieq count multieq_list =
-      match multieq_list with
-      | [] -> []
-      | hd::tl -> (Printf.sprintf "%s = %s_%s_%d ();\n" hd.A.fname ctx.A.context hd.A.fname count
-                  ) :: (gen_function_call_for_multieq (count+1) tl)
-    in String.concat "\n" (gen_function_call_for_multieq 0 ctx.A.cbody)
+  (* NOTE: this's still not working as expected... `visited` doesn't prevent duplicate assignments *)
```
Comment (jzacsh) [inline src/codegen.ml]: can you just write a bit more about what the problem and theoretical solution is. i don't understand from this comment as it is now.

Comment (nnhoang) [inline src/codegen.ml]: I meant that it's pretty ugly but we're not gonna fix it in this project since efficiency is not necessary.

**Comment by jzacsh** [comment]
looks good! please merge when you're ready!

**Pull request merged**